scalanet/src/io/iohk/scalanet/peergroup/kademlia/KNetwork.scala

```diff
@@ -39,60 +49,78 @@ object KNetwork {
+        }
+        .filterNot(_ == null)
+    }
 
+    override def findNodes(to: NodeRecord[A], request: FindNodes[A]): Task[Nodes[A]] = {
+      requestTemplate(to, request, { case n @ Nodes(_, _, _) => n })
+    }
+
+    override def ping(to: NodeRecord[A], request: Ping[A]): Task[Pong[A]] = {
+      requestTemplate(to, request, { case p @ Pong(_, _) => p })
+    }
+
+    private def requestTemplate[Request <: KRequest[A], Response <: KResponse[A]](
+        to: NodeRecord[A],
+        message: Request,
+        pf: PartialFunction[KMessage[A], Response]
+    ): Task[Response] = {
       peerGroup
         .client(to.routingAddress)
         .bracket { clientChannel =>
-          makeFindNodesRequest(message, clientChannel)
+          sendRequest(message, clientChannel, pf)
         } { clientChannel =>
           clientChannel.close()
         }
     }
 
-    override def findNodes: ConnectableObservable[(FindNodes[A], Nodes[A] => Task[Unit])] = {
-      val serverChannels: Observable[Channel[A, KMessage[A]]] = peerGroup.server().collectChannelCreated
-      try {
-        val collectedChannels = serverChannels.mergeMap { channel: Channel[A, KMessage[A]] =>
-          channel.in
-            .collect {
-              case f @ FindNodes(_, _, _) =>
-                (f, nodesTask(channel))
-            }
-        }
-        val connectableNodes = ConnectableSubject[(FindNodes[A], Nodes[A] => Task[Unit])](collectedChannels)
-        peerGroup.server().collectChannelCreated.foreach(_.in.connect())
-        peerGroup.server().connect()
-        connectableNodes
-      } finally {
-        serverChannels.map(closeIfAnError)
-      }
-    }
-    private def makeFindNodesRequest(message: FindNodes[A], clientChannel: Channel[A, KMessage[A]]): Task[Nodes[A]] = {
-      clientChannel.sendMessage(message).timeout(requestTimeout).flatMap { _ =>
-        {
-          val observable = clientChannel.in.collect { case n @ Nodes(_, _, _) => n }
-          val nodesF = observable.headL.runToFuture
-          clientChannel.in.connect()
-          Task.fromFuture(nodesF).timeout(requestTimeout)
-        }
-      }
+    private def sendRequest[Request <: KRequest[A], Response <: KResponse[A]](
+        message: Request,
+        clientChannel: Channel[A, KMessage[A]],
+        pf: PartialFunction[KMessage[A], Response]
+    ): Task[Response] = {
+      for {
+        _ <- clientChannel.sendMessage(message).timeout(requestTimeout)
+        response <- clientChannel.in.refCount
+          .collect(pf)
+          .headL
+          .timeout(requestTimeout)
+      } yield response
     }
 
-    private def closeIfAnError(
+    private def closeTheChannel[Request <: KRequest[A]](
         channel: Channel[A, KMessage[A]]
-    )(maybeError: Option[Throwable]): Task[Unit] = {
-      maybeError.fold(Task.unit)(_ => channel.close())
+    )(error: Throwable): (Request, Option[KMessage[A]] => Task[Unit]) = {
+      channel.close()
+      null
```

Comment:
What's the reason _not_ to return `null`? 

(The arguments about `null` vs, say `Option`, ignoring whether or not they are valid, refer to pubic defs because, it is claimed, it increases type safety. Within the context of an implementation, we are quite free to eschew types if we wish).

Reply:
I guess the idea is to not use `null` at all?

Reply:
> I guess the idea is to not use null at all?

I do endorse that. Furthermore, I will not approve any PR that adds a `null` reference that can be removed (and this one can).
 
> What's the reason not to return null?

If you don't use the type system, you force everyone to pay attention to your method's implementation and keep track in their head whether they should expect a null reference. Hence, even you super hidden private methods now need to be checked and memorised as safe vs unsafe when used... I will never get why people prefer to do the work in their mind that the compiler can do for free. 

> ... refer to pubic defs...

could you share any reference stating this?

Reply:
I've used `None`.